File: bt-java/src/main/java/bt/Tracker.java
package bt;

import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.time.LocalDate;
import java.time.Month;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Tracker {
    private Map<Integer, Map<Integer, Map<Type, Map<Category, List<Transaction>>>>> transactions;
    private static final String DATA = "transactions.ser";
    private static final Integer NUM_MONTHS = 12;

    public Tracker() {
        this.transactions = new HashMap<>();
        loadData();
    }

    /**
     * Adds an income to the tracker
     * @param amount        the amount of income
     * @param description   the description of the transaction
     * @param category      the category from a predefined list
     * @param date          date of the transaction as a {@link LocalDate} object
     */
    public void addIncome(double amount, String description, Category category, LocalDate date) {
        addTransaction(amount, description, category, date, Type.INCOME);
    }

    /**
     * Adds an expense to the tracker
     * @param amount        the amount spent
     * @param description   the description of the transaction
     * @param category      the category from a predefined list
     * @param date          date of the transaction as a {@link LocalDate} object
     */
    public void addExpense(double amount, String description, Category category, LocalDate date) {
        addTransaction(amount, description, category, date, Type.EXPENSE);
    }

    private void addTransaction(double amount, String description, Category category, LocalDate date, Type type) {
        Transaction newTransaction = new Transaction(amount, description, category, date, type);
        int year = date.getYear();
        int month = date.getMonthValue();
        
        transactions
            .computeIfAbsent(year, k -> new HashMap<>())
            .computeIfAbsent(month, k -> new HashMap<>())
            .computeIfAbsent(type, k -> new HashMap<>())
            .computeIfAbsent(category, k -> new ArrayList<>())
            .add(newTransaction);

        saveData();
    }

    @SuppressWarnings("unchecked")
    private void loadData() {
        try (FileInputStream fileIn = new FileInputStream(DATA);
        ObjectInputStream in = new ObjectInputStream(fileIn)) {
            transactions = (Map<Integer, Map<Integer, Map<Type, Map< Category, List<Transaction>>>>>) in.readObject();
        } catch (IOException | ClassNotFoundException e) {
            e.printStackTrace();
        }
    }
    
    private void saveData() {
        try (FileOutputStream fileOut = new FileOutputStream(DATA);
            ObjectOutputStream out = new ObjectOutputStream(fileOut)) {
                out.writeObject(transactions);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    /**
     * Displays a report of the current month
     */
    public Map<String, Integer> displayMonthlyReport() {
        return displayMonthlyReport(LocalDate.now().getMonthValue(), LocalDate.now().getYear());
    }

    /**
     * Displays a report of the specified month in the current year
     */
    public Map<String, Integer> displayMonthlyReport(Integer month) {
        return displayMonthlyReport(LocalDate.now().getMonthValue(), LocalDate.now().getYear());
    }

    /**
     * Displays a report of the specified month and year
     */
    public Map<String, Integer> displayMonthlyReport(Integer month, Integer year) {
        return displayMonthlyReportHelper(month, year, false);
    }

    private Map<String, Integer> displayMonthlyReportHelper(Integer month, Integer year, boolean concise) {
        Map<String, Integer> retMap = new HashMap<>();
        retMap.put("Income", 0);
        retMap.put("Expenses", 0);

        System.out.println("Monthly Report for " + Month.of(month).name() + " " + year);
        System.out.println("--------------------------------------------");

        Map<Integer, Map<Type, Map< Category, List<Transaction>>>> yearTransactions = transactions.get(year);
        if (yearTransactions == null) {
            System.out.println("\nNo transactions for the year " + year);
            return retMap;
        }

        Map<Type, Map<Category, List<Transaction>>> monthTransactions = yearTransactions.get(month);
        if (monthTransactions == null) {
            System.out.println("\nNo transactions for the month " + Month.of(month));
            return retMap;
        }

        // display transactions by category in chronological order
        Map<Category, List<Transaction>> incomeByCategory = monthTransactions.getOrDefault(Type.INCOME, null);
        Map<Category, List<Transaction>> expensesByCategory = monthTransactions.getOrDefault(Type.EXPENSE, null);

        int incomeTotal = 0, expenseTotal = 0;

        if (concise) {
            incomeTotal = gatherTotal(incomeByCategory);
            expenseTotal = gatherTotal(expensesByCategory);

        } else {
            System.out.println("\nINCOME:");
            System.out.println("--------------------------------------------");
            incomeTotal = displayByCategory(incomeByCategory);
            System.out.println("--------------------------------------------");
            System.out.println("\nEXPENSES:");
            System.out.println("--------------------------------------------");
            expenseTotal = displayByCategory(expensesByCategory);
            System.out.println("--------------------------------------------");
        }

        System.out.println("\nTotal income: " + incomeTotal + "\nTotal expenses: " + expenseTotal + "\nNet Income: " + (incomeTotal - expenseTotal));
        retMap.put("Income", incomeTotal);
        retMap.put("Expenses", expenseTotal);
        return retMap;
    }

    /**
     * displays transactions by category in chronological order
     * @param transactionsByCategory a map containing a list of transactions per category
     * @return the total amount processed
     */
    private int displayByCategory(Map<Category, List<Transaction>> transactionsByCategory) {
        if (transactionsByCategory == null) {
            System.out.println("No transactions found.");
            return 0;
        }

        int totalAmount = 0;
        for (Category category : transactionsByCategory.keySet()) {
            System.out.println(category.getName() + ":");
            List<Transaction> categoryTransactions = transactionsByCategory.get(category);
            if (categoryTransactions == null) {
                System.out.println("No transactions for this category");
                continue;
            }

            categoryTransactions.sort(Comparator.comparing(Transaction::getDate));
            for (Transaction transaction : categoryTransactions) {
                System.out.println(transaction.toString());
                totalAmount += transaction.getAmount();
            }
        }

        return totalAmount;
    }

    private int gatherTotal(Map<Category, List<Transaction>> transactionsByCategory) {
        int total = 0;
        if (transactionsByCategory == null) {
            return total;
        }

        for (List<Transaction> transactions : transactionsByCategory.values()) {
            for (Transaction transaction : transactions) {
                total += transaction.getAmount();
            }
        }

        return total;
    }

    /**
     * Displays a report for the current year
     */
    public void displayYearlyReport() {
        displayYearlyReport(LocalDate.now().getYear());
    }

    /**
     * Displays a report for the specified year
     */
    public void displayYearlyReport(Integer year) {
        displayYearlyReportHelper(year, false);
    }

    private void displayYearlyReportHelper(Integer year, boolean concise) {
        int totalIncome = 0;
        int totalExpenses = 0;
        for (int i = 1; i <= NUM_MONTHS; i++) {
            Map<String, Integer> totals = concise ? displayConciseMonthlyReport(i, year) : displayMonthlyReport(i, year);
            totalIncome += totals.get("Income");
            totalExpenses += totals.get("Expenses");
        }

        System.out.println("\n--------------------------------------------");
        System.out.println("--------------------------------------------");
        System.out.println("Average Income: " + (totalIncome / NUM_MONTHS)
            + "\nAverage Expenses: " + (totalExpenses / NUM_MONTHS)
            + "\nTotal Income: " + totalIncome
            + "\nTotal Expenses: " + totalExpenses
            + "\nNet Income: " + (totalIncome - totalExpenses));
    }

    /**
     * Displays a concise report for the current month
     */
    public Map<String, Integer> displayConciseMonthlyReport() {
        return displayConciseMonthlyReport(LocalDate.now().getMonthValue(), LocalDate.now().getYear());
    }

    /**
     * Displays a concise report for the specified month in the current year
     */
    public Map<String, Integer> displayConciseMonthlyReport(int month) {
        return displayConciseMonthlyReport(month, LocalDate.now().getYear());
    }

    /**
     * Displays a concise report for the specified month and year
     */
    public Map<String, Integer> displayConciseMonthlyReport(int month, int year) {
        return displayMonthlyReportHelper(month, year, true);
    }

    /**
     * Displays a concise report for the current year
    */
    public void displayConciseYearlyReport() {
        displayConciseYearlyReport(LocalDate.now().getYear());
    }

    /**
     * Displays a concise report for the specified year
    */
    public void displayConciseYearlyReport(int year) {
        displayYearlyReportHelper(year, true);
    }
}

File: bt-cpp/src/Tracker.cpp
#include "Tracker.h"

#include <iostream>
#include <algorithm>
#include <iomanip>
#include <sstream>

Tracker::Tracker() {
    loadData();
}

void Tracker::addIncome(double amount, std::string description, Category category, const std::chrono::system_clock::time_point& date) {
    addTransaction(amount, description, category, date, Type::INCOME);
}

void Tracker::addExpense(double amount, std::string description, Category category, const std::chrono::system_clock::time_point& date) {
    addTransaction(amount, description, category, date, Type::EXPENSE);
}

void Tracker::addTransaction(double amount, std::string description, Category category, const std::chrono::system_clock::time_point& date, Type type) {
    Transaction transaction(amount, description, category, date, type);
    std::time_t timet = std::chrono::system_clock::to_time_t(date);
    std::tm tm = *std::localtime(&timet);
    int year = tm.tm_year + 1900;
    int month = tm.tm_mon + 1;

    transactions[year][month][type][category].push_back(transaction);
    saveData();
}

void Tracker::loadData() {
    std::cout << "TODO" << std::endl;
}

void Tracker::saveData() {
    std::cout << "TODO" << std::endl;
}

std::unordered_map<std::string, int> Tracker::displayMonthlyReport(int month, int year) {
    return displayMonthlyReportHelper(month, year, false);
}

std::unordered_map<std::string, int> Tracker::displayConciseMonthlyReport(int month, int year) {
    return displayMonthlyReportHelper(month, year, true);
}

std::unordered_map<std::string, int> Tracker::displayMonthlyReportHelper(int month, int year, bool concise) {
    std::unordered_map<std::string, int> retMap;
    retMap["Income"] = 0;
    retMap["Expenses"] = 0;
    
    std::cout << "\nMonthly Report for " << intToMonth(month) << " " << std::to_string(year) << "\n";
    std::cout << "--------------------------------------------" << "\n";

    auto yearTransactions = transactions.find(year);
    if (yearTransactions == transactions.end()) {
        std::cout << "\nNo transactions for the year " << year << "\n";
        return retMap;
    }

    auto monthTransactions = yearTransactions->second.find(month);
    if (monthTransactions == yearTransactions->second.end()) {
        std::cout << "\nNo transactions for the month " << intToMonth(month) << "\n";
        return retMap;
    }
    
    auto incomeByCategory = monthTransactions->second.find(Type::INCOME) != monthTransactions->second.end()
                            ? monthTransactions->second.at(Type::INCOME)
                            : std::unordered_map<Category, std::vector<Transaction>>();

    auto expenseByCategory = monthTransactions->second.find(Type::EXPENSE) != monthTransactions->second.end()
                                ? monthTransactions->second.at(Type::EXPENSE)
                                : std::unordered_map<Category, std::vector<Transaction>>();
    
    int incomeTotal = 0, expenseTotal = 0;

    if (concise) {
        incomeTotal = gatherTotal(incomeByCategory);
        expenseTotal = gatherTotal(expenseByCategory);
    } else {
        std::cout << "\nINCOME:\n";
        std::cout << "--------------------------------------------\n";
        incomeTotal = displayByCategory(incomeByCategory);
        std::cout << "--------------------------------------------\n";
        std::cout << "\nEXPENSES:\n";
        std::cout << "--------------------------------------------\n";
        expenseTotal = displayByCategory(expenseByCategory);
        std::cout << "--------------------------------------------\n\n";
    }

    std::cout << "\nTotal income: " << std::to_string(incomeTotal) << "\nTotal expenses: " << std::to_string(expenseTotal) << "\nNet Income: " << std::to_string(incomeTotal - expenseTotal) << "\n\n";
    retMap["Income"] = incomeTotal;
    retMap["Expenses"] = expenseTotal;
    return retMap;
}

int Tracker::gatherTotal(std::unordered_map<Category, std::vector<Transaction>> transactionsByCategory) {
    int total = 0;
    for (const auto& [category, transactions] : transactionsByCategory) {
        for (const auto& transaction : transactions) {
            total += transaction.getAmount();
        }
    }

    return total;
}

int Tracker::displayByCategory(std::unordered_map<Category, std::vector<Transaction>> transactionsByCategory) {
    if (transactionsByCategory.empty()) {
        std::cout << "No transactions found.\n";
        return 0;
    }

    int total = 0;
    for (const auto& [category, transactions] : transactionsByCategory) {
        std::cout << categoryToString(category) << ":\n";
        if (transactions.empty()) {
            std::cout << "No transactions for this category\n";
            continue;
        }

        auto sortedTransactions = transactions;
        std::sort(sortedTransactions.begin(), sortedTransactions.end(),
            [](const Transaction& t1, const Transaction& t2) {
                return t1.getDate() > t2.getDate();
            });

        for (const auto& transaction : sortedTransactions) {
            std::cout << transaction.toString() << "\n";
            total += transaction.getAmount();
        }
    }

    return total;
}

std::string Tracker::categoryToString(Category category) {
    switch(category) {
        case Category::GROCERIES: return "Groceries";
        case Category::TRANSPORT: return "Transport";
        case Category::RENT: return "Rent";
        case Category::UTILITIES: return "Utilities";
        case Category::SHOPPING: return "Shopping";
        case Category::EDUCATION: return "Education";
        case Category::FOOD: return "Food";
        case Category::HEALTH: return "Health";
        case Category::ENTERTAINMENT: return "Entertainment";
        case Category::SALARY: return "Salary";
        case Category::INVESTMENTS: return "Investments";
        case Category::GIFTS: return "Gifts";
        case Category::INTEREST: return "Interest";
        default: return "Unknown";
    }
}

std::string Tracker::intToMonth(int month) {
    switch (month) {
        case 1:
            return "January";
        case 2:
            return "February";
        case 3:
            return "March";
        case 4:
            return "April";
        case 5:
            return "May";
        case 6:
            return "June";
        case 7:
            return "July";
        case 8:
            return "August";
        case 9:
            return "September";
        case 10:
            return "October";
        case 11:
            return "November";
        case 12:
            return "December";
    }

    return "Invalid";
}

void Tracker::displayYearlyReport(int year) {
    displayYearlyReportHelper(year, false);
}

void Tracker::displayConciseYearlyReport(int year) {
    displayYearlyReportHelper(year, true);
}

void Tracker::displayYearlyReportHelper(int year, bool concise) {
    int totalIncome = 0;
    int totalExpenses = 0;

    for (int i = 1; i <= NUM_MONTHS; i++) {
        std::unordered_map<std::string, int> totals = concise ? displayConciseMonthlyReport(i, year) : displayMonthlyReport(i, year);
        totalIncome += totals.at("Income");
        totalExpenses += totals.at("Expenses");
    }

    std::cout << "\n--------------------------------------------\n";
    std::cout << "--------------------------------------------\n\n";

    std::cout << "Average Income: " << std::to_string(totalIncome / (float) NUM_MONTHS) << "\n";
    std::cout << "Average Expenses: " << std::to_string(totalExpenses / (float) NUM_MONTHS) << "\n";
    std::cout << "Total Income: " << std::to_string(totalIncome) << "\nTotal Expenses: " << std::to_string(totalExpenses) << "\n";
    std::cout << "Net Income: " << std::to_string(totalIncome - totalExpenses) << "\n\n";
}
